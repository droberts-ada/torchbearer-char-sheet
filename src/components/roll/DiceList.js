import React from 'react';
import PropTypes from 'prop-types';

import _ from 'underscore';

import diceImages from './images/diceImages';

import './styles/DiceList.css';

const DiceList = ({ dice, name, extra }) => {
  let count = dice.length.toString();
  let successDice = "";
  if (extra || extra === 0) {
    count += ' ';
    count += extra >= 0 ? '+' : '-';
    count += ' ' + Math.abs(extra);
    count += ' = ' + (dice.length + extra);
    if (extra > 0) {
      successDice = _.times(extra, i => (
        <li
          className="die-image-container free-success"
          key={'s_' + i}
        >
          <label>S</label>
        </li>
      ))
    }
  }
  return (
    <div className="dice-list">
      <h2>
        {name + ' '}
        ({count})
      </h2>
      <ul>
        {successDice}
        {
          dice.map((die, i) => {
            let className = "die-image-container" + (die.rerolled ? " rerolled" : "");
            if (i >= dice.length + extra) {
              className += " cancelled";
            }
            return (
              <li
                className={className}
                key={die.id}
              >
                <img
                  src={diceImages[die.face]}
                  alt={`die-${die.face}`}
                />
              </li>
            );
          })
        }
      </ul>
    </div>
  );
};

DiceList.propTypes = {
  dice: PropTypes.array.isRequired,
  name: PropTypes.string.isRequired,
  extra: PropTypes.number,
};

export default DiceList;