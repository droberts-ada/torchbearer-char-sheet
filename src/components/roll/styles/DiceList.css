.dice-list h2 {
  /* text-align: left; */
}

.dice-list ul {
  display: flex;
  flex-wrap: wrap;
}

.dice-list .die-image-container img {
  height: 4rem;
}

.dice-list .die-image-container {
  height: 4rem;
  
  display: inline-block;
  margin-left: .4rem;
  margin-right: .4rem;
  margin-bottom: .25rem;
  margin-top: .25rem;

  position: relative;
}

.dice-list .die-image-container.rerolled:before {
  content: "";
  display: block;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background: rgba(0,0,0, 0.25);
  border-radius: 1rem;
}

.dice-list .die-image-container.cancelled:before {
  content: "X";
  display: block;
  position: absolute;
  top: -.6rem;
  left: 50%;
  width: 4rem;
  margin-left: -1.5rem;

  font-size: 5em;
  font-family: 'Macondo', fantasy, cursive, sans-serif;
  
  color: rgba(86.3%, 15.3%, 15.3%, 90%);
}

.dice-list .die-image-container.free-success {
  /* display: block; */
  height: 4rem;
  width: 4rem;
  flex-shrink: 0;

  box-sizing: border-box;
  border-radius: .8rem;
  border: 1px solid black;

  display: flex;
  align-items: center;
  justify-content: center;

  font-size: 2.5em;
  font-family: 'Macondo', fantasy, cursive, sans-serif;
}