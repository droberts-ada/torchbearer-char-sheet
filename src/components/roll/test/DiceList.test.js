import React from 'react';
import renderer from 'react-test-renderer';

import DiceList from '../DiceList';

describe('DiceList', () => {
  const dice = [
    {id: 0, face: 6, rerolled: false},
    {id: 1, face: 6, rerolled: false},
    {id: 2, face: 5, rerolled: false},
    {id: 3, face: 4, rerolled: false},
  ];

  it('Renders without rerolls or extra successes', () => {
    const tree = renderer
      .create(
        <DiceList
          dice={dice}
          name="Successes"
          />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders with extra successes', () => {
    const tree = renderer
      .create(
        <DiceList
          dice={dice}
          name="Successes"
          extra={2}
          />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders with negative successes', () => {
    const tree = renderer
      .create(
        <DiceList
          dice={dice}
          name="Successes"
          extra={-2}
          />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders with no dice', () => {
    const tree = renderer
      .create(
        <DiceList
          dice={[]}
          name="Successes"
          />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders with rerolls', () => {
    const diceCopy = JSON.parse(JSON.stringify(dice));
    diceCopy[0].rerolled = true;
    diceCopy[2].rerolled = true;
    const tree = renderer
      .create(
        <DiceList
          dice={diceCopy}
          name="Successes"
          />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})

