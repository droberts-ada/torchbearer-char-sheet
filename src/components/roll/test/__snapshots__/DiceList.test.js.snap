// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiceList Renders with extra successes 1`] = `
<div
  className="dice-list"
>
  <h2>
    Successes 
    (
    4 + 2 = 6
    )
  </h2>
  <ul>
    <li
      className="die-image-container free-success"
    >
      <label>
        S
      </label>
    </li>
    <li
      className="die-image-container free-success"
    >
      <label>
        S
      </label>
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-6"
        src="dice-6.png"
      />
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-6"
        src="dice-6.png"
      />
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-5"
        src="dice-5.png"
      />
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-4"
        src="dice-4.png"
      />
    </li>
  </ul>
</div>
`;

exports[`DiceList Renders with negative successes 1`] = `
<div
  className="dice-list"
>
  <h2>
    Successes 
    (
    4 - 2 = 2
    )
  </h2>
  <ul>
    
    <li
      className="die-image-container"
    >
      <img
        alt="die-6"
        src="dice-6.png"
      />
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-6"
        src="dice-6.png"
      />
    </li>
    <li
      className="die-image-container cancelled"
    >
      <img
        alt="die-5"
        src="dice-5.png"
      />
    </li>
    <li
      className="die-image-container cancelled"
    >
      <img
        alt="die-4"
        src="dice-4.png"
      />
    </li>
  </ul>
</div>
`;

exports[`DiceList Renders with no dice 1`] = `
<div
  className="dice-list"
>
  <h2>
    Successes 
    (
    0
    )
  </h2>
  <ul>
    
  </ul>
</div>
`;

exports[`DiceList Renders with rerolls 1`] = `
<div
  className="dice-list"
>
  <h2>
    Successes 
    (
    4
    )
  </h2>
  <ul>
    
    <li
      className="die-image-container rerolled"
    >
      <img
        alt="die-6"
        src="dice-6.png"
      />
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-6"
        src="dice-6.png"
      />
    </li>
    <li
      className="die-image-container rerolled"
    >
      <img
        alt="die-5"
        src="dice-5.png"
      />
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-4"
        src="dice-4.png"
      />
    </li>
  </ul>
</div>
`;

exports[`DiceList Renders without rerolls or extra successes 1`] = `
<div
  className="dice-list"
>
  <h2>
    Successes 
    (
    4
    )
  </h2>
  <ul>
    
    <li
      className="die-image-container"
    >
      <img
        alt="die-6"
        src="dice-6.png"
      />
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-6"
        src="dice-6.png"
      />
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-5"
        src="dice-5.png"
      />
    </li>
    <li
      className="die-image-container"
    >
      <img
        alt="die-4"
        src="dice-4.png"
      />
    </li>
  </ul>
</div>
`;
